File: algorithms/strings/robin-karp-rolling-hash/python/robin_karp_rolling_hash.py
def robin_karp_rolling_hash(arr):
    """
    Find first occurrence of pattern in text using rolling hash.

    Input: [text_len, ...text, pattern_len, ...pattern]
    Returns: index of first match, or -1
    """
    idx = 0
    tlen = arr[idx]; idx += 1
    text = arr[idx:idx + tlen]; idx += tlen
    plen = arr[idx]; idx += 1
    pattern = arr[idx:idx + plen]

    if plen > tlen:
        return -1

    BASE = 31
    MOD = 1000000007

    # Compute pattern hash and initial text window hash
    p_hash = 0
    t_hash = 0
    power = 1
    for i in range(plen):
        p_hash = (p_hash + (pattern[i] + 1) * power) % MOD
        t_hash = (t_hash + (text[i] + 1) * power) % MOD
        if i < plen - 1:
            power = (power * BASE) % MOD

    for i in range(tlen - plen + 1):
        if t_hash == p_hash:
            match = True
            for j in range(plen):
                if text[i + j] != pattern[j]:
                    match = False
                    break
            if match:
                return i

        if i < tlen - plen:
            t_hash = (t_hash - (text[i] + 1)) % MOD
            t_hash = (t_hash * pow(BASE, MOD - 2, MOD)) % MOD
            t_hash = (t_hash + (text[i + plen] + 1) * power) % MOD

    return -1


if __name__ == "__main__":
    print(robin_karp_rolling_hash([5, 1, 2, 3, 4, 5, 2, 1, 2]))   # 0
    print(robin_karp_rolling_hash([5, 1, 2, 3, 4, 5, 2, 3, 4]))   # 1
    print(robin_karp_rolling_hash([4, 1, 2, 3, 4, 2, 5, 6]))      # -1
    print(robin_karp_rolling_hash([4, 1, 2, 3, 4, 1, 4]))         # 3

File: algorithms/strings/robin-karp-rolling-hash/csharp/RobinKarpRollingHash.cs
using System;

public class RobinKarpRollingHash
{
    public static int Solve(int[] arr)
    {
        int idx = 0;
        int tlen = arr[idx++];
        int[] text = new int[tlen];
        for (int i = 0; i < tlen; i++) text[i] = arr[idx++];
        int plen = arr[idx++];
        int[] pattern = new int[plen];
        for (int i = 0; i < plen; i++) pattern[i] = arr[idx++];
        if (plen > tlen) return -1;

        for (int i = 0; i <= tlen - plen; i++)
        {
            bool match = true;
            for (int j = 0; j < plen; j++)
                if (text[i + j] != pattern[j]) { match = false; break; }
            if (match) return i;
        }
        return -1;
    }

    static void Main(string[] args)
    {
        Console.WriteLine(Solve(new int[] { 5, 1, 2, 3, 4, 5, 2, 1, 2 }));
        Console.WriteLine(Solve(new int[] { 5, 1, 2, 3, 4, 5, 2, 3, 4 }));
        Console.WriteLine(Solve(new int[] { 4, 1, 2, 3, 4, 2, 5, 6 }));
        Console.WriteLine(Solve(new int[] { 4, 1, 2, 3, 4, 1, 4 }));
    }
}
